bayem/vba.py

```diff
@@ -0,0 +1,553 @@
+        i_iter = 0
+        while True:
+            i_iter += 1
+
+            self.update_parameters(k, J)
+
+            self.update_noise(k, J)
+
+            for idx in self.options.index_ARD:
+                mean = self.m[idx]
+                var = 1 / self.L[idx, idx]
+                new_var = mean ** 2 + var
+                self.x0.precision[idx, idx] = 1 / new_var
+
+            logger.info(f"current mean: {self.m}")
+
+            k, J = self.p(self.m)
+
+            f_new = self.free_energy(k, J)
+
+            for i in self.options.index_ARD:
+                d = 0.5
+                r = 2 * self.x0.precision[i, i]
+                f_new -= (d - 2) * special.digamma(d)
+                f_new -= d * (1 + np.log(r))
+                f_new -= special.gammaln(d)
+                for noise in self.noise_groups:
+                    f_new += (d - 2) * np.log(self.s[noise])
+
+            logger.info(f"Free energy of iteration {i_iter} is {f_new}")
+
+            self.result.try_update(
+                f_new, self.m, self.L, self.c, self.s, self.x0.parameter_names
+            )
+            if self.stop_criteria(f_new, i_iter):
+                break
+
+        delta_f = self.f_old - f_new
+        logger.debug(f"Stopping VB. Iterations:{i_iter}, free energy change {delta_f}.")
+
+        self.result.nit = i_iter
+
+        self.result.t = perf_counter() - t0
+        self.result.param0 = self.x0
+        self.result.noise0_dict = self.noise0
+        self.result.noise_dict = self.result.noise
+        self.result.noise0 = self.p.original_noise(self.noise0)
+        self.result.noise = self.p.original_noise(self.result.noise)
+        return self.result
+
+    def update_parameters(self, k, J):
+        # fw model parameter update
+        m0, L0 = self.x0.mean, self.x0.precision
+        self.L = (
+            sum([self.s[i] * self.c[i] * J[i].T @ J[i] for i in self.noise_groups]) + L0
+        )
+        self.L_inv = np.linalg.inv(self.L)
+
+        Lm = sum(
+            [
+                self.s[i] * self.c[i] * J[i].T @ (-k[i] + J[i] @ self.m)
+                for i in self.noise_groups
+            ]
+        )
+        Lm += L0 @ m0
+        self.m = Lm @ self.L_inv
+
+    def update_noise(self, k, J):
+        # noise parameter update
+        for i in self.noise_groups:
+            try:
+                update_noise = self.options.update_noise[i]
+            except TypeError:
+                update_noise = self.options.update_noise
+
+            if not update_noise:
+                return
+
+            # if update_noise[i]:
+            # formula (30)
+            c0i, s0i = self.noise0[i].shape, self.noise0[i].scale
+            self.c[i] = len(k[i]) / 2 + c0i
+            # formula (31)
+            s_inv = (
+                1 / s0i
+                + 0.5 * k[i].T @ k[i]
+                + 0.5 * np.trace(self.L_inv @ J[i].T @ J[i])
+            )
+            self.s[i] = 1 / s_inv
+
+    def free_energy(self, k, J):
```

Comment:
Minor:
I like to have it as a separate method. Would it probably make any better sense, if we have it as a class (static) method instead (just rough idea, as I myself implemented once before this way)?

Reply:
Really good point. I left that out, because that free method would have over 10 parameters and something inside me refuses to write such a method. I suggest to leave it for now and discuss a remedy in #80. OK?